### PR TITLE
Adding tls host name extension (SNI)

### PR DESCRIPTION
**Pull request by V-Gotwig**
The SNI feature is essential in some cloud setups, e.g. when accessing mqtt broker behind a ha-proxy serving multiple domains from the same IP address.

Signed-off-by: Viktor Gotwig <viktor.gotwig@q-loud.de>

### CODE REVIEW

**Comment by karlp** [comment]
See also https://github.com/eclipse/mosquitto/pull/268 and https://github.com/eclipse/mosquitto/pull/221

File: lib/net_mosq.c [hunk 1]
```diff
@@ -4,12 +4,12 @@ Copyright (c) 2009-2014 Roger Light <roger@atchoo.org>
...
 are made available under the terms of the Eclipse Public License v1.0
 and Eclipse Distribution License v1.0 which accompany this distribution.
- 
+
 The Eclipse Public License is available at
    http://www.eclipse.org/legal/epl-v10.html
 and the Eclipse Distribution License is available at
   http://www.eclipse.org/org/documents/edl-v10.php.
- 
+
```
Comment (karlp) [inline lib/net_mosq.c]: you want to avoid making unrelated whitespace changes.

File: lib/net_mosq.c [hunk 2]
```diff
@@ -635,6 +635,14 @@ int _mosquitto_socket_connect_step3(struct mosquitto *mosq, const char *host, ui
...
 		SSL_set_bio(mosq->ssl, bio, bio);
 
+    /*
+     * required for the SNI resolving
+     */
+    if(SSL_set_tlsext_host_name(mosq->ssl, host) != 1) {
+			COMPAT_CLOSE(mosq->sock);
+			return MOSQ_ERR_TLS;
+    }
+
```
Comment (karlp) [inline lib/net_mosq.c]: Looks like you have a different whitespace setting in your editor here.  you want to maintain the same code style as the surrounding code.

Comment (V-Gotwig) [reply lib/net_mosq.c]: ok, fixed

**Comment by tedder** [comment]
Noting (for future googling) this is necessary for AWS "Amazon MQ" to work, as AWS makes heavy use of SNI. Currently it causes the generic error:
```
Error: A network protocol error occurred when communicating with the broker.
```

**Comment by V-Gotwig** [comment]
sorry, the commit b2444ec1236f8987f139a80090d92104c3f17af was wrong here, should get a separate pull request

**Comment by ralight** [comment]
I've merged this against the develop branch manually, so I'm going to close this PR. Thanks very much for your contribution.
